src/associations/has-many.js

```diff
@@ -416,24 +404,35 @@ class HasMany extends Association {
    * Un-associate one or several target rows.
    *
    * @param {Model} sourceInstance instance to un associate instances with
-   * @param {Model|Model[]|string|string[]|number|number[]} [targetInstances] Can be an Instance or its primary key, or a mixed array of instances and primary keys
+   * @param {Model|Model[]|string|string[]|number|number[]} [rawTargetInstances] Can be an Instance or its primary key, or a mixed array of instances and primary keys
    * @param {object} [options] Options passed to `target.update`
    *
    * @returns {Promise}
    */
-  async remove(sourceInstance, targetInstances, options = {}) {
-    const update = {
-      [this.foreignKey]: null,
-    };
-
-    targetInstances = this.toInstanceArray(targetInstances);
+  async remove(sourceInstance, rawTargetInstances, options = {}) {
+    const targetInstances = this.toInstanceArray(rawTargetInstances);
 
     const where = {
-      [this.foreignKey]: sourceInstance.get(this.sourceKey),
-      [this.target.primaryKeyAttribute]: targetInstances.map(targetInstance => targetInstance.get(this.target.primaryKeyAttribute)),
+      [this.target.primaryKeyAttribute]: targetInstances.map(
+        associatedObject => associatedObject[this.target.primaryKeyAttribute],
+      ),
     };
 
-    await this.target.unscoped().update(update, { ...options, where });
+    // update or delete based on the target's foreign key
+    const foreignKeyIsNullable = this.target.rawAttributes[this.foreignKey].allowNull !== false;
+    const target = this.target.unscoped();
+    if (foreignKeyIsNullable && !options.destroy) {
+      const update = {
+        [this.foreignKey]: null,
+      };
+
+      where[this.foreignKey] = sourceInstance.get(this.sourceKey);
```

Comment:
the same condition should be used for both update and destroy, this should be moved back inside `where` for both branches like it was before